python_modules/dagster/dagster_tests/utils_tests/test_data_loader.py

```diff
@@ -0,0 +1,39 @@
+import asyncio
+from typing import List
+
+from dagster._model import DagsterModel
+from dagster._utils.aiodataloader import DataLoader
+
+
+def test_basic():
+    class Thing(DagsterModel):
+        key: str
+        batch_keys: List[str]
+
+    class ThingLoader(DataLoader[str, Thing]):
+        async def batch_load_fn(self, keys: List[str]):
+            return [Thing(key=key, batch_keys=keys) for key in keys]
+
+    async def two_round_trips(loader: ThingLoader, key: str):
+        thing = await loader.load(key)
+        repeat = await loader.load(thing.key)
+        # test caching
+        assert repeat is thing
+        return thing
+
+    async def main() -> None:
+        loader = ThingLoader()
+
+        thing1, thing2, thing3 = await asyncio.gather(
+            two_round_trips(loader, "key_0"),
+            two_round_trips(loader, "key_1"),
+            two_round_trips(loader, "key_2"),
+        )
+        keys = ["key_0", "key_1", "key_2"]
+
+        # test batching
+        assert thing1.batch_keys == keys
+        assert thing2.batch_keys == keys
+        assert thing3.batch_keys == keys
+
+    asyncio.run(main())
```

Comment:
do we want anything else specific under test ?

Reply:
* Two layers of gather seems like a good idea.
* We should do some fault injection testing into the Loader. What is the behavior when you return the wrong stuff. Are the error messages good etc.
* What is the behavior when you forget to await


I think the biggest thing here is that we should _assume_ that the Loader doesn't work and try to break it before we bet the farm on this thing.